src/fromager/context.py

```diff
@@ -51,46 +52,50 @@ def __init__(
         # Push items onto the stack as we start to resolve their
         # dependencies so at the end we have a list of items that need to
         # be built in order.
-        self._build_stack = []
-        self._build_requirements = set()
+        self._build_stack: list[typing.Any] = []
+        self._build_requirements: set[tuple[NormalizedName, str]] = set()
 
         # Track requirements we've seen before so we don't resolve the
         # same dependencies over and over and so we can break cycles in
         # the dependency list. The key is the requirements spec, rather
         # than the package, in case we do have multiple rules for the same
         # package.
-        self._seen_requirements = set()
+        self._seen_requirements: set[tuple[NormalizedName, tuple[str, ...], str]] = (
+            set()
+        )
 
     @property
     def pip_wheel_server_args(self) -> list[str]:
         args = ["--index-url", self.wheel_server_url]
         parsed = urlparse(self.wheel_server_url)
-        if parsed.scheme != "https":
+        if parsed.scheme != "https" and parsed.hostname:
             args = args + ["--trusted-host", parsed.hostname]
         return args
 
-    def _resolved_key(self, req: Requirement, version: Version) -> tuple:
+    def _resolved_key(
+        self, req: Requirement, version: Version
+    ) -> tuple[NormalizedName, tuple[str, ...], str]:
         return (canonicalize_name(req.name), tuple(sorted(req.extras)), str(version))
 
-    def mark_as_seen(self, req: Requirement, version: Version):
+    def mark_as_seen(self, req: Requirement, version: Version) -> None:
         key = self._resolved_key(req, version)
         logger.debug(f"{req.name}: remembering seen sdist {key}")
         self._seen_requirements.add(key)
 
-    def has_been_seen(self, req: Requirement, version: Version) -> tuple:
+    def has_been_seen(self, req: Requirement, version: Version) -> bool:
         return self._resolved_key(req, version) in self._seen_requirements
 
     def add_to_build_order(
         self,
         req_type: str,
         req: Requirement,
         version: Version,
-        why: list,
+        why: list[typing.Any],
```

Comment:
`list[typing.Any]` isn't very useful. Does the list contain the same type of item all the time?

Reply:
I left the cleanup of `typing.Any` as a separate task cause these will most likely require type aliases (the last one on the task list in the issue)